### PR TITLE
Griddly envs

### PR DESCRIPTION
**Pull request by Bam4d**
This is an initial view of integration with griddly envs #14 :

## What it currently does:
* Only supports single agent environments
* You can add environments by just adding the appropriate yaml file to the __init__.py
https://github.com/entity-neural-network/incubator/compare/main...Bam4d:griddly_envs?expand=1#diff-aeeeac04a356f305716aafd0db2f163acdbd3b8d3f9700d78913f0c6bf111df1R10

## What it currently does not do
* no multi-agent support (theres no MA support in enn so far anyway)
* no multi-entity support (again i think we are still working on "select entity actions" #19)

## Some things I had to modify to get this to work:
* Observation and action spaces cannot sit at the class level, griddly does not know it's observation/action space until games are loaded. I had to nuke practically every case of classmethod in Environment.
https://github.com/entity-neural-network/incubator/compare/main...Bam4d:griddly_envs?expand=1#diff-4fec3bd175e5dac8bad40b56be738d715c333f98384be3592c83c8c9a4f9410aL213

## Other points I ran into
* EntityID cannot be any object. I tried to set my entity ids by their xyz coordinates as a tuple and this broke the raggedbuffer (I didnt investivate this heavily, but i think because my EntityID was iterable, something went wrong)

* There are parts of the training loop that expect at least one of each entity defined in the observation space, but many observations may only contain a subset.
https://github.com/entity-neural-network/incubator/compare/main...Bam4d:griddly_envs?expand=1#diff-6712e554e4c620bc39182daa80b4105c623e5330a8bdf24341563fc9ca350262R92

## Griddly API changes
* Move the entity observation and mask generation into a custom observer type so Its all in c++ and fast?
* Possibly Add a unique identifier to objects so I can use this in EntityID?

### CODE REVIEW

File: enn_ppo/enn_ppo/train.py
Comment (Theomat) [inline enn_ppo/enn_ppo/train.py:204]: This looks like an operation that we would do most of the times with an `EnvList` so I would create both `obs_space` and `action_space` methods in `EnvList`

File: entity_gym/entity_gym/environment.py
Comment (Theomat) [inline entity_gym/entity_gym/environment.py:174]: I'm completely in favour of moving this to an instance method rather than a class method, see #23 for additional possible issues that can be solved with this choice.

**Comment by Theomat** [comment]
Nice work! Now let's convince @cswinter that moving `obs_space`and `action_space` to instance methods is the right choice.

**Comment by Bam4d** [comment]
> Nice work! Now let's convince @cswinter that moving `obs_space`and `action_space` to instance methods is the right choice.

Definitely from a integration point of view, I'd have to hard code the obs spaces for every different Griddly environment. If its an instance method I can just generate it automatically when the GDY is loaded.

**Comment by vwxyzjn** [comment]
@Bam4d #70 was merged. Would you mind adding griddly as an optional dependency? Examples can be found here 

https://github.com/vwxyzjn/cleanrl/blob/f901415a961e3c88782ff55fb9b0f4e5f340f168/pyproject.toml#L17-L51

So you would probably do

```
poetry add griddly --optional
```

Then manually modify the `pyproject.toml` to add

```
[tool.poetry.extras]
griddly = ["griddly"]
```

Then run `poetry lock`, `poetry install -E griddly`.

File: entity_gym/entity_gym/envs/griddly/enn_wrapper.py
Comment (cswinter) [inline entity_gym/entity_gym/envs/griddly/enn_wrapper.py:100]: I'd replace this with a dict lookup, somewhat more concise and probably faster as well.

**Review by cswinter — CHANGES_REQUESTED** [review]
Very exciting, looking forward to getting this merged and running some nontrivial environments!

On the whole, this looks good to me and I left some smaller notes inline. A few higher-level comments:

**Package structure**: The `entity-gym` package should remain a lightweight dependency that only defines the environment interface, generic helpers/utilities/wrappers, and a few self-contained sample environments. We should move the griddly wrapper to its own package or a new enn-zoo package.

**Environment class methods**: The main reason why `obs_space` and `action_space` are class methods is that it should be possible to obtain the `ObsSpace`/`ActionSpace` without initialization of the environment. Initializing the environment is a potentially costly operation that may not even be possible to perform within certain runtime contexts (e.g., an environment that needs to talk to a server for setup).

I'd be happy to entertain any proposal that retains this property. Luckily, no design change is necessary in the context of this PR since we can just use the magical power of Python to fit griddly envs into the existing interface:

```python
class GriddlyEnv(Environment):
    @classmethod
    @abstractmethod
    def _griddly_env(cls) -> GymWrapper:
        pass
  
    # ...

def load_gridly_env(yaml_file: str):
    env = GymWrapper(yaml_file)
    class ConcreteGriddlyEnv(GriddlyEnv):
        def _griddly_env(cls) -> GymWrapper:
            return env
    return ConcreteGriddlyEnv
```

**Tests**: If possible, we should add a test that runs a few steps of training with a Griddly env and makes sure that at least nothing crashes, similar to https://github.com/entity-neural-network/incubator/blob/f7c9ab4140056bcf7284e26acbf68e32c35cab9a/enn_ppo/enn_ppo/test_training.py#L4. This would greatly increase our CI coverage, and ensure that nobody accidentally breaks your integration.

**Documentation**: There should be some short docs somewhere describing any setup you need to perform to run the Griddly env. Looks like it involves cloning a GitHub repo?

File: entity_gym/entity_gym/envs/griddly/enn_wrapper.py
Comment (Bam4d) [reply entity_gym/entity_gym/envs/griddly/enn_wrapper.py:100]: This will be pushed into c++ , so these inefficiencies will disappear fairly quickly

File: CONTRIBUTING.md
Comment (Bam4d) [reply CONTRIBUTING.md:20]: I had an error installing this to do with Cargo. I dont remeber the exact error message but installing rust fixed it

Comment (cswinter) [reply CONTRIBUTING.md]: OK, I raised an issue for it: https://github.com/PyO3/maturin/discussions/723

**Comment by cswinter** [comment]
@Bam4d It looks like the Rust issue might just have been related to old pip install: https://github.com/PyO3/maturin/discussions/723
I assume we won't run into this with poetry.

File: enn_zoo/pyproject.toml
Comment (cswinter) [inline enn_zoo/pyproject.toml:10]: This dependency seems to be unused? I don't see a fundamental reason why entity_zoo should be tied to a specific network architecture/implementation.

File: entity_gym/pyproject.toml
Comment (cswinter) [inline entity_gym/pyproject.toml:12]: entity_gym should not be depending on entity_zoo. I think the right thing to do here is to just import and add on to the `ENV_REGISTRY` inside griddly.

**Review by cswinter — APPROVED** [review]
Please simplify the dependency structure before merging, but otherwise this LGTM now once tests pass!

File: enn_zoo/pyproject.toml
Comment (Bam4d) [reply enn_zoo/pyproject.toml:10]: I think this was just me trying to get poetry to be less of a pain. Wasn't entirely sure what I was doing here to be honest

**Pull request merged**
